crates/nu-command/src/strings/mime_/guess.rs

```diff
@@ -0,0 +1,160 @@
+                description: "Guess the MIME types from the paths and return a table.",
+                result: Some(Value::list(
+                    vec![
+                        Value::record(
+                            record!("name" => Value::string("video.mkv".to_string(), NO_SPAN), "type" => Value::string("video/x-matroska", NO_SPAN)),
+                            NO_SPAN,
+                        ),
+                        Value::record(
+                            record!("name" => Value::string("audio.mp3".to_string(), NO_SPAN), "type" => Value::string("audio/mpeg", NO_SPAN)),
+                            NO_SPAN,
+                        ),
+                    ],
+                    NO_SPAN,
+                )),
+            },
+            Example {
+                example: r#"["mkv" "mp3"] | mime guess -e"#,
+                description: "Guess the MIME types from the extensions and return a table.",
+                result: Some(Value::list(
+                    vec![
+                        Value::record(
+                            record!("name" => Value::string("mkv".to_string(), NO_SPAN), "type" => Value::string("video/x-matroska", NO_SPAN)),
+                            NO_SPAN,
+                        ),
+                        Value::record(
+                            record!("name" => Value::string("mp3".to_string(), NO_SPAN), "type" => Value::string("audio/mpeg", NO_SPAN)),
+                            NO_SPAN,
+                        ),
+                    ],
+                    NO_SPAN,
+                )),
+            },
+            Example {
+                example: r#"let input = glob -d 1 * | wrap filename; $input | merge ($input | get filename | mime guess | select type)"#,
+                description: "Add a MIME type column to a table.",
+                result: Some(Value::list(
+                    vec![Value::record(
+                        record!("filename" => Value::string("...".to_string(), NO_SPAN), "type" => Value::string("...", NO_SPAN)),
+                        NO_SPAN,
+                    )],
+                    NO_SPAN,
+                )),
+            },
+        ]
+    }
+
+    fn run(
+        &self,
+        engine_state: &EngineState,
+        stack: &mut Stack,
+        call: &Call,
+        input: PipelineData,
+    ) -> Result<PipelineData, ShellError> {
+        let use_extension: bool = call.has_flag(engine_state, stack, "extension")?;
+
+        let guess_function: fn(&str) -> mime_guess::MimeGuess = if use_extension {
+            mime_guess::from_ext
+        } else {
+            // HACK I don't know how to satisfy the compiler here without a closure, but I cannot return the function directly.
+            // If I do, I get an error that the types are different or that a value does not live long enough when the function is called.
+            |input| mime_guess::from_path(input)
+        };
```

Comment:
I guess that is due to the generic `P: AsRef<Path>` on `from_path` compared to the concrete `&str` parameter in `from_ext`?

Reply:
Yes, the genericness of `from_path` is not playing nice with the concreteness of `from_ext`. That's the reason for the error that the types mismatch. That's why I try to coerce it into the same type with a closure. Hopefully the compiler will optimize that away.

However, another error shows up later when I use the variable:  `guess_function` does not live long enough. I'm not sure why that one is caused. Isn't it supposed to be a cheaply `Copy`-able fn pointer?